### PR TITLE
feat(listing): add fulfill listing & tests

### PR DESCRIPTION
**Pull request by kwiss**
## Description

This pull request introduces significant updates to the `Orderbook` contract, focusing on the implementation of listing order fulfillment and the enhancement of testing scenarios.

#### Key Additions:

1. **Fulfilling Listing Order Functionality:**
   - A new function, `fulfill_order`, has been added to handle the process of fulfilling orders. This function takes in `execution_info` and `signer` as parameters and follows these steps:
     - Verifies the signer.
     - Validates the existence of the order and its status.
     - Checks if the order is open and not fulfilled by the same offerer.
     - Distinguishes the order type and proceeds with the fulfillment process for listing orders.

2. **Private Function for Listing Order Fulfillment:**
   - The `_fulfill_listing_order` private function has been implemented to specifically handle listing orders. It checks the order's expiration date and updates the order status to 'Fulfilled' upon successful execution.

3. **Enhanced Test Coverage:**
   - New test scenarios have been added to validate the functionality:
     - `test_create_listing_order_and_fulfill_non_existing_order`: Ensures the system correctly handles attempts to fulfill non-existing orders.
     - `test_create_listing_order_and_fulfill`: Tests the successful creation and fulfillment of a listing order.
     - `test_create_listing_order_and_fulfill_with_same_fulfiller`: Validates that an order cannot be fulfilled by the same entity that created it.

These updates aim to streamline the order fulfillment process, particularly for listing orders, while ensuring robustness through comprehensive testing.

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature (`feat:`)
- [ ] 🐛 Bug Fix (`fix:`)
- [ ] 📝 Documentation Update (`docs:`)
- [ ] 🎨 Style (`style:`)
- [ ] 🧑‍💻 Code Refactor (`refactor:`)
- [ ] 🔥 Performance Improvements (`perf:`)
- [ ] ✅ Test (`test:`)
- [ ] 🤖 Build (`build:`)
- [ ] 🔁 CI (`ci:`)
- [ ] 📦 Chore (`chore:`)
- [ ] ⏩ Revert (`revert:`)
- [ ] 🚀 Breaking Changes (`BREAKING CHANGE:`)

## Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 Documentation
- [x] 🙅 no documentation needed

### PR Title and Description Guidelines:

- Ensure your PR title follows semantic versioning standards. This helps automate releases and changelogs.
- Use types like `feat:`, `fix:`, `chore:`, `BREAKING CHANGE:` etc. in your PR title.
- Your PR title will be used as a commit message when merging. Make sure it adheres to [Conventional Commits standards](https://www.conventionalcommits.org/).

### CODE REVIEW

File: crates/ark-contracts/arkchain/src/order/types.cairo [hunk 1]
```diff
@@ -154,14 +155,28 @@ impl Felt252TryIntoOrderStatus of TryInto<felt252, OrderStatus> {
 struct ExecutionInfo {
```
Comment (glihm) [inline crates/ark-contracts/arkchain/src/order/types.cairo:155]: Should be changed to `FulfillInfo` to be consistent with the vocabulary we've chosen.

File: crates/ark-contracts/arkchain/src/order/types.cairo [hunk 2]
```diff
@@ -154,14 +155,28 @@ impl Felt252TryIntoOrderStatus of TryInto<felt252, OrderStatus> {
...
     // The hash of the order to execute.
     order_hash: felt252,
+    // The hash of the offer order to be executed with an auction.
+    offer_hash: Option<felt252>,
```
Comment (glihm) [inline crates/ark-contracts/arkchain/src/order/types.cairo:159]: This hash may be used for other purposes than auction.
I would suggest to use something more generic like `reference_order_hash` or `linked_order_hash` or `related_order_hash` something like this. WDYT?

File: crates/ark-contracts/arkchain/src/orderbook.cairo [hunk 3]
```diff
@@ -291,19 +295,77 @@ mod orderbook {
...
+            SignerValidator::verify(execution_hash, signer);
+
+            // Check if order exists
+            let order: OrderV1 = match order_read(order_hash) {
+                Option::Some(o) => o,
+                Option::None => panic_with_felt252(orderbook_errors::ORDER_NOT_FOUND),
+            };
+
+            // Check if order offerer is not execution_info fulfiller
+            assert(order.offerer != execution_info.fulfiller, orderbook_errors::ORDER_SAME_OFFERER);
```
Comment (glihm) [inline crates/ark-contracts/arkchain/src/orderbook.cairo:310]: Not sure this is a good test. In the case of Auction, the `offerer` and the `fulfiller` **MUST** be the same. As we will also verify the signatures.

File: crates/ark-contracts/arkchain/src/orderbook.cairo [hunk 4]
```diff
@@ -291,19 +295,77 @@ mod orderbook {
...
     // *************************************************************************
     // INTERNAL FUNCTIONS
     // *************************************************************************
     #[generate_trait]
     impl InternalFunctions of InternalFunctionsTrait {
+        /// Fulfill order
+        ///
+        /// # Arguments
+        /// * `execution_info` - The execution info of the order.
+        /// * `order_type` - The order type of the order.
```
Comment (glihm) [inline crates/ark-contracts/arkchain/src/orderbook.cairo:350]:
```suggestion
        /// * `order_type` - The type of the order.
```

**Review by glihm — COMMENTED** [review]
Amazing work! Few remarks:

1. I am not that happy with the database design, you need to call `order_read` then `order_status_read` then `order_type_read`. I think it would be better to rework this in order to have two database tables. One for the order content, one for the order info, where in one read, we can have all info related to the order.

2. I do think as we've seen @kwiss that renaming right now the `execution` with `fulfill` will help understanding the flow if we want to keep `execution` for Starknet operator contract. If you do prefer to use `execution` on Orderbook side, no problem at all, we just need to clarify the nomenclature.

3. Some of the comments are saying exactly what the function does. When it's obvious, it would be better to remove the comments.

File: crates/ark-contracts/arkchain/src/orderbook.cairo (hunk 3 above)
Comment (kwiss) [reply crates/ark-contracts/arkchain/src/orderbook.cairo:310]: Wdym by 
> As we will also verify the signatures.

i think it's still important to check for difference or equality, should we do it in solis ?

File: crates/ark-contracts/arkchain/src/order/types.cairo (hunk 2 above)
Comment (kwiss) [reply crates/ark-contracts/arkchain/src/order/types.cairo:159]: related_order_hash

File: crates/ark-contracts/arkchain/src/orderbook.cairo (hunk 3 above)
Comment (kwiss) [reply crates/ark-contracts/arkchain/src/orderbook.cairo:310]: i moved it in the fulfull listing function and maybe we can add a test for the auction to check specifically that the offerer is the fulfiller

File: crates/ark-contracts/arkchain/src/orderbook.cairo [hunk 5]
```diff
@@ -291,19 +295,62 @@ mod orderbook {
...
+            let status = match order_status_read(order_hash) {
+                Option::Some(s) => s,
+                Option::None => panic_with_felt252(orderbook_errors::ORDER_NOT_FOUND),
+            };
+            assert(status == OrderStatus::Open, orderbook_errors::ORDER_NOT_FULFILLABLE);
+            let order_type = match order_type_read(order_hash) {
+                Option::Some(s) => s,
+                Option::None => panic_with_felt252(orderbook_errors::ORDER_NOT_FOUND),
+            };
+            let token_hash = order.compute_token_hash();
```
Comment (remiroyc) [inline crates/ark-contracts/arkchain/src/orderbook.cairo:315]: `token_hash` seems not to be used and `order` is only used to compute the `token_hash`

File: crates/ark-contracts/arkchain/tests/common/setup.cairo [hunk 6]
```diff
@@ -237,9 +238,11 @@ fn setup_auction_order(start_price: felt252, end_price: felt252) -> (OrderV1, fe
...
 }
 
-
-fn setup(block_timestamp: u64) -> (OrderV1, Signer, felt252, felt252) {
-    let end_date = block_timestamp + (30 * 24 * 60 * 60);
+fn setup(block_timestamp: u64, isExpired: bool) -> (OrderV1, Signer, felt252, felt252) {
+    let mut end_date = block_timestamp + (30 * 24 * 60 * 60);
+    if (isExpired) {
+        end_date = block_timestamp;
+    }
```
Comment (remiroyc) [inline crates/ark-contracts/arkchain/tests/common/setup.cairo:245]:
```suggestion
fn setup(block_timestamp: u64, is_expired: bool) -> (OrderV1, Signer, felt252, felt252) {
    let mut end_date = block_timestamp + (30 * 24 * 60 * 60);
    if (is_expired) {
        end_date = block_timestamp;
    }
```

File: crates/ark-contracts/arkchain/src/orderbook.cairo (hunk 5 above)
Comment (kwiss) [reply crates/ark-contracts/arkchain/src/orderbook.cairo:315]: true

**Review by remiroyc — APPROVED** [review]

**Review by remiroyc — APPROVED** [review]

**Pull request merged**